### PR TITLE
chore: upgrade protoc used in CI

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

protoc was upgraded a while back and the new version is not compatible with the old.
`make -C proto check` started fails locally when using the new version. This updates CI to use a newer version (24)
https://github.com/protocolbuffers/protobuf/releases

## Test Plan

```
hbp ➜ d/determined git:(upgrade-protoc) ✗ make -C proto dep_check
make: Entering directory '/Users/hmd/projects/da/determined/proto'
echo "libprotoc 3.24.54" | tee /dev/stderr | \
	perl -ne '(/libprotoc (?:3?\.)?(\d+)/ && $1>=24) || \
		(print("Please install protoc version >= 24\n") && exit 1)'
libprotoc 3.24.54
make: Leaving directory '/Users/hmd/projects/da/determined/proto'
hbp ➜ d/determined git:(upgrade-protoc) ✗ make -C proto dep_check
make: Entering directory '/Users/hmd/projects/da/determined/proto'
echo "libprotoc 3.23.54" | tee /dev/stderr | \
	perl -ne '(/libprotoc (?:3?\.)?(\d+)/ && $1>=24) || \
		(print("Please install protoc version >= 24\n") && exit 1)'
libprotoc 3.23.54
Please install protoc version >= 24
make: *** [Makefile:37: dep_check] Error 1
make: Leaving directory '/Users/hmd/projects/da/determined/proto'
hbp ➜ d/determined git:(upgrade-protoc) ✗ make -C proto dep_check
make: Entering directory '/Users/hmd/projects/da/determined/proto'
echo "libprotoc 23.54" | tee /dev/stderr | \
	perl -ne '(/libprotoc (?:3?\.)?(\d+)/ && $1>=24) || \
		(print("Please install protoc version >= 24\n") && exit 1)'
libprotoc 23.54
Please install protoc version >= 24
make: *** [Makefile:37: dep_check] Error 1
make: Leaving directory '/Users/hmd/projects/da/determined/proto'
hbp ➜ d/determined git:(upgrade-protoc) ✗ make -C proto dep_check
make: Entering directory '/Users/hmd/projects/da/determined/proto'
echo "libprotoc 24.54" | tee /dev/stderr | \
	perl -ne '(/libprotoc (?:3?\.)?(\d+)/ && $1>=24) || \
		(print("Please install protoc version >= 24\n") && exit 1)'
libprotoc 24.54
make: Leaving directory '/Users/hmd/projects/da/determined/proto'
hbp ➜ d/determined git:(upgrade-protoc) ✗
```

## Commentary (optional)

https://hpe-aiatscale.slack.com/archives/C02PV33GSN5/p1712089871140569

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by dzhu** [comment]
Will this require everyone to upgrade their local version?

**Comment by hamidzr** [comment]
> Will this require everyone to upgrade their local version?

yes. We had a talk about this in the backend team around when it's okay to require dependency upgrades and the consensus was that as long as it's communicated we're good with it and we didn't go into more details. What do you think

**Review by dzhu — APPROVED** [review]
Sure, just checking that it was considered at some point.

**Comment by hamidzr** [comment]
I wanna update https://github.com/determined-ai/determined/blob/ed7a283fccbdf56f8156686fc224183298c29e40/CONTRIBUTING.md#L58 but I haven't been able to pin down which specific version introduces the material change.

**Review by salonig23 — APPROVED** [review]

**Comment by hamidzr** [comment]
@varlogtim had set a cutoff here at >v22 so maybe we could expand it to that but since we're testing 24.3 with the ci and it's been out for a while let's just go for that?

File: proto/Makefile [hunk 1]
```diff
@@ -35,8 +35,8 @@ get-deps:
 
 build/proto.stamp: $(source_files)
 	protoc --version | tee /dev/stderr | \
-		perl -ne '(/libprotoc (?:3?\.)?(\d+)\.(?:\d+)/ && $$1>=15 && $$1<22) || \
-			(print("Please install protoc where version >= 3.15 and < 3.22\n") && exit 1)'
+		perl -ne '(/libprotoc (?:24\.)?(\d+)\.(\d+)/ && $$1>=24 && ($$1>24 || $$2>=3)) || \
```
Comment (erikwilson) [inline proto/Makefile:38]: It looks like this regex change may not be doing what is intended. At some point protoc dropped the major version from the version number (altho `v3` tags are still being created). The `?:` part of the regex indicates a non-matching group, so the major part should not change it is just the minor and patch versions being assigned to $1 and $2. The max minor version check was dropped, does protobuf v26 also work here?

Comment (erikwilson) [reply proto/Makefile:38]: More info on the versioning change:
https://protobuf.dev/news/2022-05-06/, basically the major version is reserved for language specific api updates, but previously protoc would report the major version as 3. It should probably stay as 3 or >=3.

We should also make sure whatever minor version is easily installable for devs. Right now we can `brew install protobuf@21`, and `brew install protobuf` currently installs v26.

Comment (hamidzr) [reply proto/Makefile]: I'm thinking to leave it a >21 to keep it simple and the min version similar to ci?

Comment (erikwilson) [reply proto/Makefile]: personally I would prefer stability to running the latest thing, unless there is a particular reason we need to upgrade.  eventually something will be mismatched between versions, causing breaking CI issues again.

Comment (hamidzr) [reply proto/Makefile]: made sure we keep allowing > 3.24.x

**Comment by hamidzr** [comment]
makes sense to wait for the unforking to happen before landing this

**Pull request merged**